tests/functional/src/test/java/org/mule/functional/classloading/isolation/DefaultClassPathClassifierTestCase.java

```diff
@@ -0,0 +1,199 @@
+
+    when(context.getDependencyGraph()).thenReturn(dependencyGraph);
+    when(context.getExtensionBasePackages()).thenReturn(Lists.newArrayList("org.mule.invalid"));
+    when(context.getRootArtifactClassesFolder()).thenReturn(rootArtifactClassesFolder);
+    when(rootArtifactClassesFolder.exists()).thenReturn(false);
+
+    Properties serviceProperties = new Properties();
+    serviceProperties.setProperty(DefaultClassPathClassifier.SERVICE_PROVIDER_CLASS_NAME, ECHO_SERVICE_CLASS);
+    File metaInfFolder = new File(folder.getRoot(), "META-INF");
+    assertThat(metaInfFolder.mkdir(), is(true));
+
+    serviceProperties
+        .store(new FileWriter(new File(metaInfFolder, DefaultClassPathClassifier.SERVICE_PROPERTIES_FILE_NAME)),
+               "Service properties for discover process");
+
+    when(context.getClassPathClassLoader()).thenReturn(new URLClassLoader(new URL[] {folder.getRoot().toURI().toURL()}, null));
+
+  }
+
+  /**
+   * Validates the logic to discover extensions by scanning locations of service.properties files.
+   */
+  @Test
+  public void testServiceDiscoverProcess() {
+    MavenArtifact serviceArtifact = new MavenArtifact.MavenArtifactBuilder().withGroupId("org.mule").withArtifactId("service")
+        .withType("jar").withScope("provided").build();
+
+    String artifactPath = folder.getRoot().getParentFile().getParentFile().getAbsolutePath() + separator;
+    when(mapping.getArtifactId(artifactPath)).thenReturn(serviceArtifact.getArtifactId());
+    when(context.getMavenMultiModuleArtifactMapping()).thenReturn(mapping);
+    when(context.getServicesExclusion()).thenReturn(Lists.newArrayList());
+
+    File serviceGroupIdFolder = new File(folder.getRoot(), serviceArtifact.getGroupId());
+    assertThat(serviceGroupIdFolder.mkdir(), is(true));
+    File serviceArtifactIdFolder = new File(serviceGroupIdFolder, serviceArtifact.getArtifactId());
+    assertThat(serviceArtifactIdFolder.mkdir(), is(true));
+    when(context.getClassPathURLs()).thenReturn(Lists.newArrayList());
+
+    MavenArtifact rootArtifact = new MavenArtifact.MavenArtifactBuilder().withGroupId("org.mule").withArtifactId("service-test")
+        .withType("jar").withScope("compile").build();
+    when(context.getDependencyGraph().getRootArtifact()).thenReturn(rootArtifact);
+    when(context.getDependencyGraph().getDependencies()).thenReturn(Sets.newHashSet());
+
+    ArtifactUrlClassification classification = classPathClassifier.classify(context);
+    assertThat(classification.getServiceUrlClassifications().size(), is(1));
+    ServiceUrlClassification serviceUrlClassification = classification.getServiceUrlClassifications().get(0);
+    // No need to test this logic as it is already tested in unit tests for dependency resolver code
+    assertThat(serviceUrlClassification.getUrls().size(), is(0));
+    assertThat(serviceUrlClassification.getName(), equalTo(ECHO_SERVICE_CLASS));
+
+    verify(context, times(8)).getDependencyGraph();
```

Comment:
Could you add named constants for these numbers? I can't deduce from the tests whether a 8 makes sense or not